bluemira/equilibria/opt_constraint_funcs.py

```diff
@@ -0,0 +1,315 @@
+        Number of PF coils
+    n_CS: int
+        Number of CS coils
+    PF_Fz_max: float
+        Maximum vertical force on each PF coil [MN]
+    CS_Fz_sum_max: float
+        Maximum total vertical force on the CS stack [MN]
+    CS_Fz_sep_max: float
+        Maximum vertical separation force in the CS stack [MN]
+    scale: float
+        Current scale with which to calculate the constraints
+
+    Returns
+    -------
+    constraint: np.ndarray
+        Updated constraint vector
+    """
+    n_coils = len(vector)
+    currents = scale * vector
+
+    # get coil force and jacobian
+    F = np.zeros((n_coils, 2))
+
+    for i in range(2):  # coil force
+        # NOTE: * Hadamard matrix product
+        F[:, i] = currents * (a_mat[:, :, i] @ currents + b_vec[:, i])
+
+    F /= scale  # Scale down to MN
+
+    # Absolute vertical force constraint on PF coils
+    constraint[:n_PF] = F[:n_PF, 1] ** 2 - PF_Fz_max**2
+
+    if n_CS != 0:
+        # vertical forces on CS coils
+        cs_fz = F[n_PF:, 1]
+        # vertical force on CS stack
+        cs_z_sum = np.sum(cs_fz)
+        # Absolute sum of vertical force constraint on entire CS stack
+        constraint[n_PF] = cs_z_sum**2 - CS_Fz_sum_max**2
+        for i in range(n_CS - 1):  # evaluate each gap in CS stack
+            # CS seperation constraints
+            f_sep = np.sum(cs_fz[: i + 1]) - np.sum(cs_fz[i + 1 :])
+            constraint[n_PF + 1 + i] = f_sep - CS_Fz_sep_max
+
+    # calculate constraint jacobian
+    if grad.size > 0:
+        dF = np.zeros((n_coils, n_coils, 2))  # noqa :N803
+        im = currents.reshape(-1, 1) @ np.ones((1, n_coils))  # current matrix
+        for i in range(2):
+            dF[:, :, i] = im * a_mat[:, :, i]
+            diag = (
+                a_mat[:, :, i] @ currents
+                + currents * np.diag(a_mat[:, :, i])
+                + b_vec[:, i]
+            )
+            np.fill_diagonal(dF[:, :, i], diag)
+
+        # Absolute vertical force constraint on PF coils
+        grad[:n_PF] = 2 * dF[:n_PF, :, 1]
+
+        if n_CS != 0:
+            # Absolute sum of vertical force constraint on entire CS stack
+            grad[n_PF] = 2 * np.sum(dF[n_PF:, :, 1], axis=0)
+
+            for i in range(n_CS - 1):  # evaluate each gap in CS stack
+                # CS separation constraint Jacobians
+                f_up = np.sum(dF[n_PF : n_PF + i + 1, :, 1], axis=0)
+                f_down = np.sum(dF[n_PF + i + 1 :, :, 1], axis=0)
+                grad[n_PF + 1 + i] = f_up - f_down
+    return constraint
+
+
+def coil_field_constraints(
+    constraint, vector, grad, ax_mat, az_mat, bxp_vec, bzp_vec, B_max, scale
+):
+    """
+    Current optimisation poloidal field constraints on coils
+
+    Parameters
+    ----------
+    constraint: np.ndarray
+        Constraint array (modified in place)
+    vector: np.ndarray
+        Current vector
+    grad: np.ndarray
+        Constraint Jacobian (modified in place)
+    a_mat: np.ndarray
```

Comment:
Parameters in docstring should match arguments.  What do`b_x_vec` and `b_z_vec` represent here physically when they are nonzero? Is the maximum field B_max not sufficient?